### PR TITLE
masync: fix memory leaks and coverity issues

### PR DESCRIPTION
**Pull request by blazej-smorawski**
This change is [Reviewable](https://reviewable.io/reviews/pmem/miniasync/37)

### CODE REVIEW

File: src/vdm.c [hunk 1]
```diff
@@ -12,14 +12,17 @@ struct vdm {
...
 /*
- * vdm_new -- returns NULL if failed to allocate memory for struct vdm
- * or vdm_data_init failed
+ * vdm_new -- returns NULL if descriptor is NULL, failed to allocate memory
+ * for struct vdm or vdm_data_init failed.
  */
 struct vdm *
 vdm_new(struct vdm_descriptor *descriptor)
 {
+	if (descriptor == NULL)
```
Comment (pbalcer) [inline src/vdm.c:21]: No, this type of check is almost never a good idea. Errors should be caught at the earliest possible moment. This NULL-check should be resolved at the call site.

File: tests/memcpy_threads/memcpy_threads.c [hunk 2]
```diff
@@ -5,17 +5,20 @@
 #include <string.h>
 #include <time.h>
 #include "libminiasync.h"
+#include "core/os.h"
 
 int
 test_threads_memcpy_multiple_same_char(unsigned memcpy_count,
```
Comment (pbalcer) [inline tests/memcpy_threads/memcpy_threads.c:11]: this function and the one below are almost identical. Try to abstract away the boilerplate.

File: tests/memcpy_threads/memcpy_threads.c [hunk 3]
```diff
@@ -56,11 +60,24 @@ test_threads_memcpy_multiple_same_char(unsigned memcpy_count,
...
-		if (memcmp(sources[i], destinations[i], sizes[i]) !=
-			0) {
+		if (memcmp(sources[i], destinations[i], sizes[i]) != 0) {
 			fprintf(stderr,
 				"Memcpy nr. %u result is wrong! "
 				"Returning\n", i);
+
+			/* Cleanup */
+			for (unsigned j = 0; j < memcpy_count * n; j++) {
+				free(sources[j]);
```
Comment (pbalcer) [inline tests/memcpy_threads/memcpy_threads.c:70]: use MALLOC/FREE in tests.

File: tests/memcpy_threads/memcpy_threads.c [hunk 4]
```diff
@@ -56,11 +60,24 @@ test_threads_memcpy_multiple_same_char(unsigned memcpy_count,
...
 
 	for (unsigned i = 0; i < memcpy_count * n; i++) {
-		if (memcmp(sources[i], destinations[i], sizes[i]) !=
-			0) {
+		if (memcmp(sources[i], destinations[i], sizes[i]) != 0) {
 			fprintf(stderr,
 				"Memcpy nr. %u result is wrong! "
 				"Returning\n", i);
+
+			/* Cleanup */
```
Comment (pbalcer) [inline tests/memcpy_threads/memcpy_threads.c:68]:
```
ret = 1;
goto cleanup;
```
?

**Review by wlemkows — COMMENTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/37)* status: all files reviewed, 4 unresolved discussions (waiting on @blazej-smorawski)

**Review by blazej-smorawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/37)* status: 1 of 4 files reviewed, 4 unresolved discussions (waiting on @pbalcer and @wlemkows)

___
*[src/vdm.c, line 21 at r1](https://reviewable.io/reviews/pmem/miniasync/37#-MuPdxbHLeh7fnOzbeRg-r1-21:-MuQ1kns4eklRV1UiRgs:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/915ce1887e02e45a67d4c5683ebc859299bd3623/src/vdm.c#L21)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

No, this type of check is almost never a good idea. Errors should be caught at the earliest possible moment. This NULL-check should be resolved at the call site.
</blockquote></details>

Done.
___
*[tests/memcpy_threads/memcpy_threads.c, line 11 at r1](https://reviewable.io/reviews/pmem/miniasync/37#-MuPdxbHLeh7fnOzbeRi-r1-11:-MuQ1leN5RF3hXegm48n:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/915ce1887e02e45a67d4c5683ebc859299bd3623/tests/memcpy_threads/memcpy_threads.c#L11)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

this function and the one below are almost identical. Try to abstract away the boilerplate.
</blockquote></details>

Done.
___
*[tests/memcpy_threads/memcpy_threads.c, line 68 at r1](https://reviewable.io/reviews/pmem/miniasync/37#-MuPdxbHLeh7fnOzbeRi-r1-68:-MuQ1mSWCALGN9R0_zVc:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/915ce1887e02e45a67d4c5683ebc859299bd3623/tests/memcpy_threads/memcpy_threads.c#L68)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

```
ret = 1;
goto cleanup;
```
?
</blockquote></details>

Done.
___
*[tests/memcpy_threads/memcpy_threads.c, line 70 at r1](https://reviewable.io/reviews/pmem/miniasync/37#-MuPdxbHLeh7fnOzbeRi-r1-70:-MuQ1xNy1J-AZJa4L3j9:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/915ce1887e02e45a67d4c5683ebc859299bd3623/tests/memcpy_threads/memcpy_threads.c#L70)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

use MALLOC/FREE in tests.
</blockquote></details>

Done.

**Review by pbalcer — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/37)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @blazej-smorawski, @pbalcer, and @wlemkows)

___
*[tests/memcpy_threads/memcpy_threads.c, line 13 at r2](https://reviewable.io/reviews/pmem/miniasync/37#-MuVb-Z93Yj2LO8Nnp3k:-MuVb-Z93Yj2LO8Nnp3l:b-fxhrje) ([raw file](https://github.com/pmem/miniasync/blob/426052b87f92c337f8c6ca0906e2a16926941281/tests/memcpy_threads/memcpy_threads.c#L13)):*
> ```C
>  * test_threads_memcpy_multiple -- test multiple memcpy operations
>  * on a specific descriptor.
>  * test_type = 0 => sequence
> ```

use an enum

**Comment by kswiecicki** [comment]
___
*[src/vdm.c, line 2 at r2](https://reviewable.io/reviews/pmem/miniasync/37#-MuVua5dCHigbp4AH51A:-MuVua5dCHigbp4AH51B:b-aya5f3) ([raw file](https://github.com/pmem/miniasync/blob/426052b87f92c337f8c6ca0906e2a16926941281/src/vdm.c#L2)):*
> ```C
> // SPDX-License-Identifier: BSD-3-Clause
> /* Copyright 2021, Intel Corporation */
> ```

2022.
I'm not sure how did this pass license check.



_Code quote:_
```C
 2021
```

**Review by blazej-smorawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/37)* status: 1 of 4 files reviewed, 1 unresolved discussion (waiting on @kswiecicki, @pbalcer, and @wlemkows)

___
*[src/vdm.c, line 2 at r2](https://reviewable.io/reviews/pmem/miniasync/37#-MuVua5dCHigbp4AH51A:-Muj6hCM8y7k8K22yRnX:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/426052b87f92c337f8c6ca0906e2a16926941281/src/vdm.c#L2)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

2022.
I'm not sure how did this pass license check.
</blockquote></details>

Done.
___
*[tests/memcpy_threads/memcpy_threads.c, line 13 at r2](https://reviewable.io/reviews/pmem/miniasync/37#-MuVb-Z93Yj2LO8Nnp3k:-Muj6jBTAqfToQT08XPQ:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/426052b87f92c337f8c6ca0906e2a16926941281/tests/memcpy_threads/memcpy_threads.c#L13)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

use an enum
</blockquote></details>

Done.

**Review by pbalcer — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/37#-:-MujMtWx0cmFYSZsl-2X:bnfp4nl)* status: 1 of 4 files reviewed, all discussions resolved (waiting on @kswiecicki, @pbalcer, and @wlemkows)

**Comment by kswiecicki** [comment]
:lgtm:

**Pull request merged**